src/main/clojure/clara/rules/durability/fressian.clj

```diff
@@ -235,6 +235,16 @@
                 (read [_ rdr tag component-count]
                   (read-with-meta rdr #(apply list %))))}}
 
+   "clj/emptylist"
+   {:class clojure.lang.PersistentList$EmptyList
```

Comment:
This is a sneaky case. The case for `clojure.lang.ASeq` is meant to catch the Clojure collection types not otherwise specified here, but  `clojure.lang.PersistentList$EmptyList` is a special inner class that apparently only implements the interfaces, which differs from how the mos to the other Clojure data structures are implemented.

However, it is concerning. Classes like this are really "implementation details" of Clojure. I don't know that there is a good way around this though, other than to keep it consistent with Clara's supported versions of Clojure.

One, potentially obvious, sounding fix for this sort of issue would be to implement readers and writers against the interfaces instead of these concrete classes. However, this can lead to issues with multiple handlers matching for some objects, which will fail. This is similar to the multiple inheritance sort of issues that happen with protocol extensions. Many concrete collection types will implement many common interfaces. It tends to cause problems quickly. The recommendation I've seen is to only target extending concrete classes instead of interfaces. (Technically the "lookup" mechanism of Fressian is extensible and could be used to implement something similar to multimethod "prefers", but that still gets complicated and brittle.)

The downside, is that we have to be sure to cover all possible types that Clara may be serializing in its own network and working memory structures.

I don't see a better way right now other than doing what you did here and just adding the classes. This made me look around in Clojure to recall if there are any other special classes used to represent "empty" collections. It doesn't see to be the case for vectors, sets, or maps. Clojure eventually backed off an earlier [approach to using different classes for certain "smaller" collection sizes](https://dev.clojure.org/jira/browse/CLJ-1517) too, so that shouldn't be a problem.

Reply:
Yeah, as far as I can tell there is no consistent convention or API in Clojure that would allow us to serialize and deserialize these collections while maintaining their concrete class without special-casing each concrete type.  It is admittedly unfortunate that something like [Java's collection constructor](https://docs.oracle.com/javase/8/docs/api/java/util/ArrayList.html#ArrayList-java.util.Collection-) isn't available for Clojure's data structures.